### PR TITLE
Notification stack

### LINKED ISSUE #83: Notification toasts are displayed on top of each other
**Description**
When several notification toasts are displayed, they are overlaying each other

**Steps To Reproduce**

Add clear steps on how to reproduce the bug

1. Go to Multix at 'https://multix.chainsafe.io/'
2. Initiate a balance transfer
3. When prompted to sign the extrinsic with the extension, cancel it. This will create a toast with the status "error". Toasts error are set to not auto-disappear, this is fine https://github.com/ChainSafe/Multix/blob/867fe5a5f52ea1f3477ce261eb8ef567048c6eaa/ui/src/components/ToastBar.tsx#L29-L30
4. Now if you don't click on the small cross to close it, any other toast will appear on top of it.

Bellow is a screenshot with a long error toast, and a "Tx broadcasted" toast on top of it.

**Expected Behaviour**
the toasts should ideally stack on top of each other.

**Screenshots / Video**
image

### PR DESCRIPTION
**Pull request by Tbaut**
Based on https://github.com/ChainSafe/Multix/pull/113 from @dj-pundir

closes #83

### CODE REVIEW

**Comment by Tbaut** [comment]
We need @ksingh-vertex to sign the CLA before this can be merged

**Comment by dj-pundir** [comment]
> We need @ksingh-vertex to sign the CLA before this can be merged

CLA sign: Done. Thanks!!

**Comment by asnaith** [comment]
@Tbaut thanks for the tips on how to test it easily! That was a big time saver. 

This is looking great @dj-pundir

Screenshot 2023-05-15 at 1 50 55 PM

**Review by asnaith — APPROVED** [review]

File: packages/ui/src/components/ToastBar.tsx [hunk 1]
```diff
@@ -28,8 +31,8 @@ const ToastBar = ({ toast }: Props) => {
   return (
     <Snackbar
```
Comment (Lykhoyda) [inline packages/ui/src/components/ToastBar.tsx:32]: We are rendering Snackbar inside MuiSnackbar is it expected?

File: packages/ui/src/components/Snackbar.tsx [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import { Stack, Snackbar as MuiSnackbar, styled } from '@mui/material'
+import { useToasts } from '../contexts/ToastContext'
+import ToastBar from './ToastBar'
```
Comment (Lykhoyda) [inline packages/ui/src/components/Snackbar.tsx:3]: We can group all the files ToastBar, ToastContent and Snackbar in one folder as it's a one component

File: packages/ui/src/components/ToastBar.tsx [hunk 3]
```diff
@@ -42,15 +45,7 @@ const ToastBar = ({ toast }: Props) => {
...
-          title={title}
-          key={id}
-          id={id}
-          link={link}
-        />
-      }
+      message={<ToastContent toast={toast} />}
     />
   )
 }
```
Comment (Lykhoyda) [inline packages/ui/src/components/ToastBar.tsx:51]: we can remove the wrapping part as we have no styles

export default styled(ToastBar)(
  ({ theme }) => `
`
)

File: packages/ui/src/components/ToastBar.tsx (hunk 1 above)
Comment (Tbaut) [reply packages/ui/src/components/ToastBar.tsx:32]: Yes this is a weird construct, `Snackbar > Stack > Snackbar`, this also looks weird to me, but works without any additional lib (the `notistack` one) so I closed my eyes on it. But maybe/probably the first Snackbar is actually not useful.

File: packages/ui/src/contexts/ToastContext.tsx [hunk 4]
```diff
@@ -1,78 +1,55 @@
...
 
-const ToastContext = React.createContext<IToastContext | undefined>(undefined)
+const ToastContext = createContext<ToastContextProps | undefined>(undefined)
 
-const ToastContextProvider = ({ children }: ToastContextProps) => {
-  const [toastQueue, setToastQueue] = useState<Toast[]>([])
-  // using useRef instead of useState to keep a tracker over the exact toast array
-  const toasts = useRef<Toast[]>([])
+const ToastProvider: React.FC<React.PropsWithChildren> = ({ children }) => {
+  const [toasts, setToasts] = useState<Toast[]>([])
```
Comment (Lykhoyda) [inline packages/ui/src/contexts/ToastContext.tsx:25]: as an alternative to not handling the state on our side, we can use https://www.npmjs.com/package/react-toastify
But it's up to us if we want to support additional code

**Review by Lykhoyda — CHANGES_REQUESTED** [review]
Some minor changes, but overall is good to merge

File: packages/ui/src/contexts/ToastContext.tsx (hunk 4 above)
Comment (Tbaut) [reply packages/ui/src/contexts/ToastContext.tsx:25]: I think it's simple enough for us to handle this directly for now at least.

File: packages/ui/src/components/ToastBar.tsx (hunk 1 above)
Comment (Tbaut) [reply packages/ui/src/components/ToastBar.tsx:32]: I managed to remove a Snackbar.

File: packages/ui/src/components/Snackbar.tsx (hunk 2 above)
Comment (Tbaut) [reply packages/ui/src/components/Snackbar.tsx:3]: done

**Pull request merged**